File: nodejs-typescript/src/domain/TransferId.ts
import { createHash } from 'node:crypto';
import type { AccountNumber } from './AccountNumber.js';
import { InvalidTransferInstructionError } from './errors.js';
import type { Money } from './Money.js';

export class TransferId {
  private constructor(readonly value: string) {}

  static derive(from: AccountNumber, to: AccountNumber, amount: Money): TransferId {
    const key = `${from.toString()}|${to.toString()}|${amount.format()}`;
    const hash = createHash('sha256').update(key, 'utf8').digest('hex');
    return new TransferId(hash);
  }

  static parse(value: string): TransferId {
    if (!value.trim()) {
      throw new InvalidTransferInstructionError('Transfer id cannot be blank');
    }
    return new TransferId(value.trim());
  }

  equals(other: TransferId): boolean {
    return this.value === other.value;
  }

  toString(): string {
    return this.value;
  }
}

File: dotnet/src/MableBank.Domain/TransferId.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using MableBank.Domain.Errors;

namespace MableBank.Domain;

public sealed record TransferId(string Value)
{
    public static TransferId Derive(AccountNumber from, AccountNumber to, Money amount)
    {
        var key = string.Join(
            '|',
            from.Value,
            to.Value,
            amount.Amount.ToString("F2", CultureInfo.InvariantCulture));
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(key));
        return new TransferId(Convert.ToHexString(hash));
    }

    public static TransferId Parse(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new InvalidTransferInstructionException("Transfer id cannot be blank");
        }

        return new TransferId(value.Trim());
    }
}
